### ISSUE TITLE
0.8 remove control messages

### ISSUE THREAD

**Issue by ChillerDragon** [teeworlds/teeworlds#3245]
Having system, game and totally different control messages adds complexity to the protocol. And I do not see the benefit of it.
All the control messages keepalive, connect, accept, close and token could just become regular system messages.
Then the control flag can be removed from the packet header. And the parsing of a packet becomes cleaner.

As of right now its something like:
```
if packet.header.flags.control {
   // assume uncompressed
   // assume only one chunk which is control even if num chunks is set to 0
} else {
   for chunk in chunks {
      if chunk.sys {
         // sys chunk
      } else {
          // game chunk
      }
   }
}
```

It could become just this:
```
for chunk in chunks {
  if chunk.sys {
    // sys chunk
    // includes all control messages
  } else {
    // game chunk
  }
}
```

The control messages would grow by 2-3 bytes if they have a chunk header. But that should barley affect anything. If someone is concerned by that overhead I am sure we can save it else where. For example the first 2 ctrl messages the client sends are full of null bytes for reflection attacks anyways. This null byte payload could be reduced by 3 bytes and then those messages are not longer. If a disconnect packet grows from 8 bytes to 11 bytes that does not seem to big of a problem.

**Comment by heinrich5991** [comment]
I think it's mostly a separation of concerns. The upper level protocol doesn't need to know about the lower level parts and vice versa, making the code more modular.

If anything, the 0.8 protocol should probably be encrypted and authenticated, e.g. QUIC. See also [RFC 7258](https://www.rfc-editor.org/rfc/rfc7258).

**Comment by ChillerDragon** [comment]
Modular as in the connect/disconnect/keepalive could be replaced without touching the gameplay affecting protocol? So if there was QUIC there would not be any control messages anymore?

**Comment by heinrich5991** [comment]
Yes. See e.g. https://github.com/ddnet/ddnet/pull/6961 which does that. There are no Teeworlds-specific control messages there anymore, it's just QUIC's protocol.

**Comment by ChillerDragon** [comment]
What about vital chunks and sequence numbers? They are part of the chunk header right now. Isn't that something which should be covered by a different abstraction layer as well? Resending lost packets. And how modular was it really when implementing QUIC? Was it smooth to swap out the control messages?

**Comment by heinrich5991** [comment]
They're all also part of the TCP part, and in they'll all become part of the QUIC implementation in the proof of concept.

> And how modular was it really when implementing QUIC? Was it smooth to swap out the control messages?

Replacing not only the control messages but also the chunks and vital message was quite smooth with QUIC.

**Comment by heinrich5991** [comment]
The close message could potentially be moved to the game protocol, since it might need to be translated at some point.
